### ISSUE TITLE
Multicapabilities retries all tests, and uses all maxInstances on fail.

### ISSUE THREAD

**Issue by eyphka** [NickTomlin/protractor-flake#18]
As an example if you pass 

multiCapabilities = [
  {
    browserName: 'chrome',
    specs: ['./someOkTest.js'],
    maxInstances: 1
  },
  {
    browserName: 'chrome',
    specs: ['./someBadTest.js'],
    maxInstances: 1
  }
];
This will run all of the tests from someBadTests, in 2 webdrivers.

And if you have multiple specs in a multicapabilities, it will run all of the specs if one fails.

**Comment by NickTomlin** [comment]
Hi there! Thanks for letting me know about this. 

So to be clear, in your example if `someOkTest` passes but `someBadTest` fails, it will re-run both `someOkTest` and `someBadTest`?

Any chance you can put together a small example repo or a PR? I'm pretty slammed at the moment but this is definitely something that would be good to fix.

**Comment by el-davo** [comment]
Just to weigh in here. I have seen this happen also when you have 10 tests suites and 1 fails it re-runs all 10. This doesnt happen when the tests fail directly, but if an error happens that is not specifically related to a test. For example if you have a plugin that is writing test results you get an error sometimes that it cannot create a directory

[chrome #01-2] PID: 19636
[chrome #01-2] Specs: C:\GIT\bla\test\e2e\integration\reference\edit\referenceEdit.spec.js
[chrome #01-2]
[chrome #01-2] Using the selenium server at http://10.73.71.214:4444/wd/hub
[chrome #01-2] Started
[chrome #01-2] .Error: ENOENT: no such file or directory, open 'C:\GIT\bla\test_results\e2e\screenshots\screenshots\9ac5a10b95
f71c7c211d7ebfd5510ab3.png'

Not sure if this is right but it is what i am observing

**Comment by NickTomlin** [comment]
@el-davo 

This is because protractor-flake does a very simple scraping of the output looking for the stacktraces of an assertion error. There's no way to connect the `Error: ENOENT: no such file or directory, open 'C:\GIT\bla\test_results\e2e\screenshots\screenshots\9ac5a10b95
f71c7c211d7ebfd5510ab3.png'` error to a spec from that stack trace so it re-runs them all. Not super great 😄  but i'm also not sure how to connect the two at the moment. Would love any ideas!

**Comment by FinKingma** [comment]
I'm also experiencing a similar problem. I ran protractor flake with 2 separate spec files, deliberately creating an error in one of them. But protractor flake reruns both javascript files.
It would help a lot if protractor flake only re-ran the files that contained errors

Re-running tests: test attempt 3
Re-running the following test files:
---/tests/referral/supplyPageSpec.js (<-- file containing spec that failed)

[launcher] Running 2 instances of WebDriver

## ...

[internet explorer 11 #1-1] PID: 29692
[internet explorer 11 #1-1] Specs: ---/tests/referral/supplyPagePreviewSpec.js (<-- file containing specs that passed last time).
[internet explorer 11 #1-1]

[EDIT]

I had a different issue. I specified the testset as environment variable and used specs instead of suites in my protractor.conf. 

This does not work with the current setup of protractor-flake.
To fix this I had to change 3 things:
1) Use suites instead of specs (suites get filtered out with the rerunning step, so they don't get run twice)
2) Set suite:null in my protractor.conf so that by default no suites are run.
3) Specifying the test suite in cmd instead of using environment var

Now only the failed files get rerun.

**Comment by NickTomlin** [comment]
This should be fixed with the new multitest support released in `2.x` (thanks to @FinKingma for adding the parser). You should be able to use the `--parser multi` option.

**Comment by patelpayal** [comment]
@NickTomlin @FinKingmaI tried protractr-flake(v2.0.1) so that we can retry running failed specs. I provided 2 specs in protractor config file under capabilities config.(1 spec always fails.)[https://github.com/angular/protractor/blob/master/docs/referenceConf.js#L168]  I am not using `suites` or `specs` in my config as all my specs are under capability section.  

protractor-flake is able to recognize the failed spec as it is printing list of failed specs. But behaviour is really odd, Number of webdriver instance it launches are 2 instead of 1. And it runs failed test 3 times + passing spec(1 time). 

Any pointers? 

Here is my log: https://gist.github.com/patelpayal/f9c906da595cf3c92862c4222ca38e08

**Comment by NickTomlin** [comment]
@patelpayal can you share your protractor config as well?

**Comment by patelpayal** [comment]
@NickTomlin 
We generate config automatically via custom lib but here is the config I am printing from protactor. I hope, its okay. https://gist.github.com/patelpayal/1c710e09abdbfb47f5154577381b7755

**Comment by FinKingma** [comment]
@NickTomlin I don't think the recent fix made any change here. It was focused on fixing issue 21 with timeouts.
This multi capabilities problem lies somewhere else. I believe this issue should be reopened.

@patelpayal 
Testfiles can be specified in a global level in your protractor config, and in the multiCapabilities array, which I see in your config patelpayal.
as in:

```
{
   specs: [not specified in your config]
   multiCapabilities: [{
       specs: [Object],
       testFileName: 'homepage.js'
   }, {
       specs: [Object],
       testFileName: 'mytest1.js'
   }]
}
```

protractor-flake will only overwrite the specs specified on the global level.
So in your case, when the tests are rerun, the conf will look something like: 

```
{
   specs: 'mytest1.js'   //because this failed and protractor-flake will have added this
   multiCapabilities: [{
       specs: [Object],
       testFileName: 'homepage.js'
   }, {
       specs: [Object],
       testFileName: 'mytest1.js'
   }]
}
```

So even while rerunning, two sets of capabilities will be run:
set 1:
- mytest1.js (from global)
- homepage.js (from multi capabilities set)

set 2:
- mytest1.js (from global)
- homepage.js (from multi capabilities set)

This is why 4 specs are executed while rerunning.

If it is possible, I would advise only specifying your test specs on global level. Protractor-flake can not overwrite testspecs specified on multi capabilities level yet. Unless you have a better idea to achieve this @NickTomlin ?

**Comment by NickTomlin** [comment]
@FinKingma 

Thanks for investigating. Yep; this is true. We deal with the top level `specs` not specs within multicapabilities. 

We could try to override this at a lower level but I can see that getting complex

**Comment by wswebcreation** [comment]
To give you more info. Protractor + Cucumber can also work with multicapabilities. I'll check next week if I can make a PR for the Cucumber parser. Will let you know.

**Comment by NickTomlin** [comment]
@wswebcreation awesome, that would be great. Thanks!

**Comment by wswebcreation** [comment]
@NickTomlin 

I see that the CucumberJS parser that was made in this [PR](https://github.com/NickTomlin/protractor-flake/pull/9) isn't working due to a newer version of CucumberJS. 
Do I need to make it "backwards compatible"?
Shall I also take that into my PR or do you want a separate PR for it, or do you prefer the parser to be standard and multi in 1?

**Comment by wswebcreation** [comment]
@NickTomlin,

Can you check this [feature](https://github.com/wswebcreation/protractor-flake/tree/fix/cucumber-multicapabilities). I'm not sure this is what you mean.

**Comment by NickTomlin** [comment]
@wswebcreation thanks! I'll try to take a look in the next day or so.

**Comment by NickTomlin** [comment]
@wswebcreation interesting! I don't really understand cucumber very well but that seems to make sense. Do you want to go ahead and PR it?

**Comment by wswebcreation** [comment]
@NickTomlin 
Can this issue be closed now, or are there still issues with multicapabilities?

**Comment by joelfogue** [comment]
@wswebcreation I'm not sure that this issue is resolved yet; I recently set up my protractor flake to run in a multicapabilities environment. So my tests run in parallel to speed up test execution and save time. Right now, protractor flake will rerun both the failing and passing tests in parallel eventhough we want ONLY the failed ones to rerun.

The desired behavior should be the following: 
1. I run 10 cucumberjs tests in parallel i.e each tests is run only once to speed up the test execution and save time. 
2. If 3 tests fail; Protractor flake should ONLY rerun the 3 that failed and do so in PARALLEL, NOT rerun both failed and passed tests in parallel.
3. IF this is already resolved could you please show us your **_protractor.conf.js_** and the **_protractor-flake command_** you ran to achieve this @wswebcreation ?

Below is mine: 

```
exports.config = {
  framework: 'custom',
  // path relative to the current config file
  frameworkPath: require.resolve('./node_modules/protractor-cucumber-framework'),

  multiCapabilities: [
    {
      'browserName': 'chrome',
      shardTestFiles: true,
      maxInstances: 6,
    }
  ],
//More config stuff here....//
}
```

And my protractor flake command: 
`protractor-flake --parser standard --node-bin node --max-attempts=3 -- protractor.conf.js`

Thanks @wswebcreation and @NickTomlin  for looking into this; again, Protractor flake works beautifully when running only a single browser instance; but as soon as your try to parallelize with multiple browsers, you lose the benefit of parallelization since protractor-flake will rerun both failed and passed tests **_--max-attempts_** times.

**Comment by wswebcreation** [comment]
Can you please add the logging of the testoutput? It seems like that the testoutput can't find any failed specs.
When it can find failed specs it should also put them in the log.

Like to hear from you

**Comment by joelfogue** [comment]
@wswebcreation Thanks for your quick response, please give me 10mins, I'll set up some dummy test files, run them with the above commands and paste the test logs here

**Comment by wswebcreation** [comment]
Tnx

Hope to check it this evening, else tomorrow

**Comment by joelfogue** [comment]
BTW are you able to achieve that desired behavior in your setup? If yes, could you maybe share your protractor.conf.js file and your protractor-flake command?

**Comment by wswebcreation** [comment]
I'm using cucumber and as an argument for the parser I now use `cucumberMulti'

You should use `multi` in stead of `standard`

Which framework do you use?

**Comment by joelfogue** [comment]
@wswebcreation First off, I'm a newbie at this e2e tests so learning as much as I can; my background is in Java.  To answer your question, I'm using cucumberJs, Protractor and Protractor-flake to handle the rerunning of failed tests; that worked great when running a single browser instance but in parallel it doesn't.
What do you mean by you should use 'multi' instead of 'standard'?  You mean cucumberMulti?

**Comment by wswebcreation** [comment]
Then please do this
- Make sure you have the latest version, should be at least 2.1.0
- Read [this](https://github.com/NickTomlin/protractor-flake/blob/master/docs/cucumber.md)

Then it should all come together  😉

**Comment by joelfogue** [comment]
I can't use Ruby, required to use Javascript thus CucumberJs. Is cucumberMulti a Ruby implementation? Could you please provide a link to cucumberMulti? Again, we can only use cucumberJs, the Javsacript implementation of cucumber. so no Ruby. 

Here are my versions: 

```
cucumberJs v1.3.0
Protractor v4.0.1
Are these the right versions to use for Protractor-flake to work?
```

I'll take a look at the link you provided, thanks

**Comment by wswebcreation** [comment]
`cucumberMulti` is the new parser for cucumberJS provided in the latest version of `protractor-flake`

Your command should be `protractor-flake --parser cucumberMulti --node-bin node --max-attempts=3 -- protractor.conf.js`

Hope both documentation files should be clear enough for you.

**Comment by joelfogue** [comment]
Reading [that link](https://github.com/NickTomlin/protractor-flake/blob/master/docs/cucumber.md) I got a better understanding or so I think, could you please confirm that I'm understanding this corrently? 
1. Because I have cucumberJs v1.3.0 (**_cucumberjs > v0.9.0 as specified in that link you provided above_**), I'm automatically using that Multi parser and NOT the standard one. 
2. Ok so I'm using the Multi-parser, I'll try that protractor-flake you provided, thanks btw. Where in the documentation do you see that you should swap out 'standard' for 'cucumberMulti'; please this will help other readers on here as I personally wasn't able to find that in the link you provided, maybe there's another documentation somewhere that we're missing ? 

Again thanks a great deal for your help so far; will try the above command right away and update you, please answer questions above if you wouldn't mind!

**Comment by wswebcreation** [comment]
1. No, you should provide the parser in your command
2. The README.md in the root is for how to implement and use (first few lines of text), the link for more info about cucumberJS usage

Good luck with it.

**Comment by joelfogue** [comment]
@wswebcreation Wonderful wealth of information, I'm sure you've helped a ton of people by going through my questions, thanks again. This should get me there and I will be back soon with whether or not it worked! 

Can you please answer this last question for me, 

This whole setup is to get protractor-flake to rerun failed tests when in parallel mode i.e you're running your tests across multiple browsers simultaneously right? Just want to clarify that because I have no problem when _rerunning_ failed tests in a single browser.  

Please confirm that this whole discussion has been about getting the rerunning of failed tests to work in multi-capabilities mode(Protractor) i.e when running tests in parallel :) 

Kind regards

**Comment by wswebcreation** [comment]
`protractor-flake` provides an array of failed `specs` and not failed testcases. That's not possible.

The failed `specs` (for cucumberJS feature-files with 1 or more scenario's) are run on all instances (browsers/devices) provided in the `multicapabilities` of the `protractor.conf.js`.

**Comment by joelfogue** [comment]
Thanks @wswebcreation , just want to add that the (_**[See example logging](https://github.com/NickTomlin/protractor-flake/blob/master/docs/cucumber.md#cucumbermulti)**_) link under the cucumberMulti heading is broken; could you please update that link when you get a chance @NickTomlin ?

**Comment by NickTomlin** [comment]
@wswebcreation I think this can be fixed by just adding `../` to the beginning of that link. Is there a specific case that is best or should we just link to the root of the cucumber fixtures?

**Comment by joelfogue** [comment]
@wswebcreation @NickTomlin Ok so I get everything working fine in parallel with that cucumberMulti parser ONLY when the --max-attemps flag is set to 2; anything greater than that reruns both the passed and failed files when only the failed should run.

I added the following the hook per the documentation to ensure that it would only rerun the failed ones but still doesn't work if --max-attempts is greater than 2

```
var afterHook = function () {
  this.After(function (scenario, callback) {
    if (scenario.isFailed()) {
      console.log('Specs failed ? :', scenario.getUri());
    }
    callback();
  });
};

module.exports = afterHook;
```

I guess --max-attemps = 2 will do for now; anything greater than 2 won't work.

**Comment by joelfogue** [comment]
@NickTomlin @wswebcreation Actually here's what seems to be happening: 
- I have 3 feature files, 2 passing tests, 1 failing test
- 3 browser instances running in parallel
- --max-attemps = 4 
- My protractor.conf.js file
  
  ```
  multiCapabilities: [
  {
    'browserName': 'chrome',
    shardTestFiles: true,
    maxInstances: 3,
  }
  ],
  ```

**On 1st attempt/run:**

```
1st feature runs //PASS
2nd feature runs //PASS
3rd feature runs  // failing feature
//The 3 features above all run in parallel 
```

**On 2nd run:**
Only the 3rd feature RERUNs as expected

**On 3rd run** since --max-attempt was set to 4, all 3 feature files re-run, both passing and failing features when ONLY the failing feature should have rerun
so we get: 

```
1st feature runs //PASS
2nd feature runs //PASS
3rd feature runs  // failing feature
```

However **on 4th run**, correct behavior is observed, only the failing feature re-runs 

**_I'll guess that the correct behavior is observed on all even-number attempt so on 2nd and 4th run, I had the correct behavior i.e only the failing feature was rerun but on odd-number attempt (3rd run) ALL were rerun, both failing and passing when only failing features should have rerun_**.

But again, if you limit **--max-attempt = 2** everything works as intended; it's just sometimes the flakiness can occur at the 2 attempt so I like to set --max-attempt to 3 or even 4 but again, you run into the trouble described above once you set **--max-attempt** to anything greater than 2

**Comment by NickTomlin** [comment]
@Joel-fogue interesting, could you provide more detail on the output between the 2nd and 3rd run? 

protractor-flake should only re-run everything if it can't find any single failing test cases in the output, so maybe there's a test that's failing in a non specific way

**Comment by wswebcreation** [comment]
@Joel-fogue ,

I've been rethinking this part and I think it all comes up to the table in the [cucumber docs](https://github.com/NickTomlin/protractor-flake/blob/master/docs/cucumber.md).

In your first run:
- multiple featurefiles + shardTestFiles : true + failure  => specs printed in log => protractor-flake will deliver the failed feature file to protractor again
  In your second run:
- 1 single feature file + shardTestFiles: true  + failure => specs are **NOT**  printed in the log => protractor-flake will deliver NO failed featurefiles and this will be printed in the testoutput `Tests failed but no specs were found. All specs will be run again`
  In the third run:
- multiple featurefiles + shardTestFiles : true + failure  => specs printed in log => protractor-flake will deliver the failed feature file to protractor again
  In your fourth run:
- 1 single feature file + shardTestFiles: true  + failure => specs are **NOT**  printed in the log => protractor-flake will deliver NO failed featurefiles and this will be printed in the testoutput `Tests failed but no specs were found. All specs will be run again`

What you need to do is to add this to your project

```
// for example called: afterHook.js
var afterHook = function () {
  this.After(function (scenario, callback) {
        if (scenario.isFailed()) {
            console.log('Specs:', scenario.getUri());
        }
        callback();
  });
};

module.exports = afterHook;
```

Then add this to you `protractor.conf.js` so it will read the `afterHook.js` file

```
cucumberOpts: {
    require: [
        'path/to/your/afterHooks.js'
    ]
}
```

This will make sure that the specs are always printed, even for the second run that, if you will check your testoutput, default has no failed specs printed.

I don't know if this is a problem from CucumberJS or from protractor (not printing the failed specs the second and the fourth time), but this is the workaround. 
I'll also need to rethink the documentation of Cucumber, I think it needs to be more "clear" then it is now.

Hope this will help you

**Comment by joelfogue** [comment]
@NickTomlin @wswebcreation Thanks to both for your quick responses; 

@wswebcreation You missed my last 2 responses; I have been using that **_afterHook.js_** and including in my **_cucumberOpts_** block inside the **_protractor.conf.js file_**, please see my previous last two comments from yesterday. 

@NickTomlin I cannot provided any stack-trace of my company's work but I will put together a dummy example in the next few days and share it with you. 

Last thing I will leave you with, just in case you hear from someone else, is I also just noticed that even when the palatalization WORKS i.e all tests pass, and the rerun tests also pass, the cucumber report doesn't reflect the correct count of all tests that were run. 

So, you have 10 tests, 8 passes, 2 are rerun right? On the rerun, those last 2 finally pass, a JSON is generated from the tests results and sent to the cucumber report plugin inside our bamboo remote agent and that cucumber reporter will only show the last 2 tests that passed; so you will see this: 

```
[chrome #01-2] 5 scenarios (5 passed)
13-Sep-2016 10:16:46    [chrome #01-2] 2 steps (2passed)
13-Sep-2016 10:16:46    [chrome #01-2] 5m55.889s
13-Sep-2016 10:16:46    [chrome #01-2] json file location: C:\Users\C\bamboo-agent-home\xml-data\build-dir\MAR-M\e2e\out\reports\cucumber-test-results.json
13-Sep-2016 10:16:46    [chrome #01-2] [10:16:45] I/local - Shutting down selenium standalone server.
13-Sep-2016 10:16:46    
13-Sep-2016 10:16:46    [10:16:46] I/testLogger - 
13-Sep-2016 10:16:46    
13-Sep-2016 10:16:46    [10:16:46] I/launcher - 0 instance(s) of WebDriver still running
13-Sep-2016 10:16:46    [10:16:46] I/launcher - chrome #01-0 passed
13-Sep-2016 10:16:46    [10:16:46] I/launcher - chrome #01-1 failed 3 test(s)
13-Sep-2016 10:16:46    [10:16:46] I/launcher - chrome #01-3 passed
13-Sep-2016 10:16:46    [10:16:46] I/launcher - chrome #01-2 passed
13-Sep-2016 10:16:46    [10:16:46] I/launcher - overall: 3 failed spec(s)
13-Sep-2016 10:16:46    [10:16:46] E/launcher - Process exited with error code 1
13-Sep-2016 10:16:49    [10:16:49] Using gulpfile C:\Users\C\bamboo-agent-home\xml-data\build-dir\MAR-YATP-M\gulpfile.js
13-Sep-2016 10:16:49    [10:16:49] Starting 'xml'...
13-Sep-2016 10:16:49    
13-Sep-2016 10:16:49    **Converting the CucumberJS JSON to XML...
13-Sep-2016 10:16:49    
13-Sep-2016 10:16:49    
13-Sep-2016 10:16:49    
13-Sep-2016 10:16:49    [10:16:49] Finished 'xml' after 349 ms
```

So it only reports those tests that were rerun. Again this could be a problem on my end but wanted to put that out there just in case it re-surfaces again from another user. Just trying to help make this tool better. 

I will keep you posted with any progress made on my end, most likely in a couple of days, thank you for this tool!

**Comment by wswebcreation** [comment]
@NickTomlin , is there an other way how I can contact you? You've got my email, can you please PM me? I've got some questions about this part.
Tnx in advance

**Comment by NickTomlin** [comment]
@wswebcreation I'm going to be busy for the next day or so but i'll email you to start the thread

**Comment by wswebcreation** [comment]
@Joel-fogue ,

Can you please update to version 2.1.2 (released today) and try again. I've refactored the `cucumberMulti` parser and with some help of @NickTomlin we changed the `index.js` file.

**Comment by joelfogue** [comment]
@wswebcreation Wow you guys are on fire! I appreciate both of your efforts on this. I'll grab a copy and try again; will update you before noon today, possibly earlier, thanks again so much; this comes at a time very much needed.

**Comment by joelfogue** [comment]
@wswebcreation  so I just installed protractor-flake and it grabbed the latest. However, I'm getting that error message below (**_`UNMET PEER DEPENDENCY protractor@>= 2.x`_**) - I understand that it means that protractor-flakes needs `a dependency of protractor whose version should be greater than 2.x.`

So I checked my protractor version with `protractor --version`and it gave me `Version 4.0.1`
I then reinstalled protractor-flake globally with `npm install -g protractor-flake` and still got error below, any idea?

```
C:\Users\<user>\AppData\Roaming\npm\protractor-flake -> C:\Users\<user>\AppData\Roaming\npm\node_modules\protractor-flake\bin\protractor-flake  
C:\Users\<user>\AppData\Roaming\npm                                                                                                                 
+-- UNMET PEER DEPENDENCY protractor@>= 2.x                                                                                                             
`-- protractor-flake@2.1.2                                                                                                                              

npm WARN protractor-flake@2.1.2 requires a peer of protractor@>= 2.x but none was installed.                                                            
```

**Comment by wswebcreation** [comment]
@Joel-fogue, 
Is protractor also installed global? The log you've pasted is only warning you, not blocking. See the readme to give `protractor-flake` the path of protractor.

For the future, it's better to keep this thread releated to the issue and check the manual / contact us.

**Comment by joelfogue** [comment]
@wswebcreation Definitely, it's installed globally that's what that `-g` flag meant in the note I sent earlier. Understood, will try to keep questions to the main issue at hand :)

What I will do if I still run into the same problem is, I will create a public repo with dummy feature files and share the repo here so you guys can clone and reproduce the same issues.

Will get back to you hopefully before noon, truly appreciate your hardwork on this.

Side note: as this tool gets better, maybe you could send a PR to the protractor team and have them integrate this into core protractor, just a thought.

**Comment by joelfogue** [comment]
Beautiful ! Works as expected.
I used the cucumberMulti parser as seen below
 **`protractor-flake --parser cucumberMulti --node-bin node --max-attempts=3 -- protractor.conf.js`**

And ran 10 chrome instances in parallel as seen in config below: 

```
  multiCapabilities: [
    {
      'browserName': 'chrome',
      maxInstances: 10,
      shardTestFiles: true,
    },
  ],
```

And what I observed this time is that protractor-flake will only rerun failed tests no matter what --max-attempt is set to.

This is exactly what I was looking for; great work @wswebcreation  and @NickTomlin 

Hopefully this project gets integrated with Protractor core as protractor can't rerun failed tests as of 09/19/2016.

I'm truly thankful for your efforts and this tool.

**Comment by wswebcreation** [comment]
Tnx for the reply!

Spread the word 👍🏼

**Comment by joelfogue** [comment]
Definitely I will; I think this issue can be close whenever you get a chance unless someone else runs into problems then they can reopen it. I already answered several questions on stackoverflow by referring them back to this github page.

**Comment by NickTomlin** [comment]
Glad to hear it worked out! Thanks!

**Comment by ammargalal** [comment]
Can we run it on jenkins only failed test cases
